internal/controller/serverclaim_controller.go

```diff
@@ -251,6 +261,81 @@ func (r *ServerClaimReconciler) removeBootConfigRefFromServerAndPowerOff(ctx con
 	return nil
 }
 
+func (r *ServerClaimReconciler) getClaimableServer(ctx context.Context, log logr.Logger, claim *metalv1alpha1.ServerClaim) (*metalv1alpha1.Server, error) {
+	var (
+		server *metalv1alpha1.Server
+		err    error
+	)
+	switch {
+	case claim.Spec.ServerRef != nil:
+		server, err = r.serverByReference(ctx, log, claim)
+	case claim.Spec.ServerSelector != nil:
+		server, err = r.serverBySelector(ctx, log, claim)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if server == nil {
+		return nil, nil
+	}
+	// Check the server state and only proceed if it is Available or Reserved
+	if server.Status.State != metalv1alpha1.ServerStateAvailable && server.Status.State != metalv1alpha1.ServerStateReserved {
+		log.V(1).Info("Server not in a claimable state", "Server", server.Name, "ServerState", server.Status.State)
+		return nil, nil
+	}
+	return server, nil
+}
+
+func (r *ServerClaimReconciler) serverByReference(ctx context.Context, log logr.Logger, claim *metalv1alpha1.ServerClaim) (*metalv1alpha1.Server, error) {
+	server := &metalv1alpha1.Server{}
+	if err := r.Get(ctx, client.ObjectKey{Name: claim.Spec.ServerRef.Name}, server); err != nil {
+		return nil, err
+	}
+	if claimRef := server.Spec.ServerClaimRef; claimRef != nil && claimRef.UID != claim.UID {
+		log.V(1).Info("Server claim ref UID does not match claim", "Server", server.Name, "ClaimUID", claimRef.UID)
+		return nil, nil
+	}
+	if claim.Spec.ServerSelector == nil {
+		return server, nil
+	}
+	selector, err := metav1.LabelSelectorAsSelector(claim.Spec.ServerSelector)
+	if err != nil {
+		return nil, err
+	}
+	if !selector.Matches(labels.Set(server.ObjectMeta.Labels)) {
+		log.V(1).Info("Specified server does not match label selector", "Server", server.Name, "Claim", claim.Name)
+		return nil, nil
+	}
+	return server, nil
+}
+
+func (r *ServerClaimReconciler) serverBySelector(ctx context.Context, log logr.Logger, claim *metalv1alpha1.ServerClaim) (*metalv1alpha1.Server, error) {
+	selector, err := metav1.LabelSelectorAsSelector(claim.Spec.ServerSelector)
+	if err != nil {
+		return nil, err
+	}
+	opts := client.ListOptions{
+		LabelSelector: selector,
+	}
+	serverList := &metalv1alpha1.ServerList{}
+	if err := r.List(ctx, serverList, &opts); err != nil {
```

Comment:
Please use a `MatchLabels` in the `List` call. That way you don't have to loop over the result set later.

Reply:
Used `client.MatchingLabelsSelector` instead of `client.MatchingLabels`. The "matcher" was chosen due to the fact, that `client.MatchingLabels` is an alias to the `map[string]string` type, however label selector might also contain `MatchExpressions` field, which is a list of more complex structs. Thus `client.MatchingLabelsSelector`, which has nested `labels.Selector` field would fit better to filter out fitting server objects.